driver/custom_endpoint_monitor.cc

```diff
@@ -0,0 +1,175 @@
+                      endpoint_identifier.c_str(), region.c_str(), custom_endpoints.size(),
+                      this->get_endpoints_as_string(custom_endpoints).c_str());
+
+          std::this_thread::sleep_for(std::chrono::nanoseconds(this->refresh_rate_nanos));
+          continue;
+        }
+        const std::shared_ptr<CUSTOM_ENDPOINT_INFO> endpoint_info =
+          CUSTOM_ENDPOINT_INFO::from_db_cluster_endpoint(custom_endpoints[0]);
+        const std::shared_ptr<CUSTOM_ENDPOINT_INFO> cache_endpoint_info =
+          custom_endpoint_cache.get(this->custom_endpoint_host_info->get_host(), nullptr);
+
+        if (cache_endpoint_info != nullptr && cache_endpoint_info == endpoint_info) {
+          const long long elapsed_time =
+            std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() - start).count();
+          std::this_thread::sleep_for(
+            std::chrono::nanoseconds(std::max(static_cast<long long>(0), this->refresh_rate_nanos - elapsed_time)));
+          continue;
+        }
+
+        MYLOG_TRACE(this->logger, 0, "Detected change in custom endpoint info for '%s':\n{%s}",
+                    custom_endpoint_host_info->get_host().c_str(), endpoint_info->to_string().c_str());
+
+        // The custom endpoint info has changed, so we need to update the set of allowed/blocked hosts.
+        std::shared_ptr<ALLOWED_AND_BLOCKED_HOSTS> allowed_and_blocked_hosts;
+        if (endpoint_info->get_member_list_type() == STATIC_LIST) {
+          allowed_and_blocked_hosts =
+            std::make_shared<ALLOWED_AND_BLOCKED_HOSTS>(endpoint_info->get_static_members(), std::set<std::string>());
+        } else {
+          allowed_and_blocked_hosts =
+            std::make_shared<ALLOWED_AND_BLOCKED_HOSTS>(std::set<std::string>(), endpoint_info->get_excluded_members());
+        }
+
+        custom_endpoint_cache.put(this->custom_endpoint_host_info->get_host(), endpoint_info,
+                                  CUSTOM_ENDPOINT_INFO_EXPIRATION_NANOS);
+        const long long elapsed_time =
+          std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() - start).count();
+        std::this_thread::sleep_for(
+          std::chrono::nanoseconds(std::max(static_cast<long long>(0), this->refresh_rate_nanos - elapsed_time)));
+      }
+
+    } catch (const std::exception& e) {
+      // Log and continue monitoring.
+      if (this->enable_logging) {
+        MYLOG_TRACE(this->logger, 0, "Error while monitoring custom endpoint: %s", e.what());
+      }
+    }
+  });
+}
+
+std::string CUSTOM_ENDPOINT_MONITOR::get_endpoints_as_string(
+  const std::vector<Aws::RDS::Model::DBClusterEndpoint>& custom_endpoints) {
+  std::string endpoints("[");
+
+  for (auto const& e : custom_endpoints) {
+    endpoints += e.GetDBClusterEndpointIdentifier();
+    endpoints += ",";
+  }
+  if (endpoints.empty()) {
+    endpoints = "<no endpoints>";
+  } else {
```

Comment:
Do you also want to do a `pop_back()` on empty case or add an ending `]`? Otherwise you would get something like `[<no endpoints>`

Reply:
`endpoints` gets overwritten if it is empty. So `[<no endpoints>` won't be an issue. However, I just realized endpoints will never be empty because it was initialized with `[`, I will fix that.